### ISSUE TITLE
Avoid embedded details about deployment environment into client build

### ISSUE THREAD

**Issue by eecavanna** [microbiomedata/nmdc-edge#88]
One of the reasons I have found it difficult to containerize this project is that the build process for the web app client involves embedding some details about the eventual deployment environment, into the build. Unlike the web app server, the built web app client cannot access the external server environment at run time (e.g. to get an updated environment variable value). That couples the build to a single deployment environment.

I want to decouple them. Part of that involves:
- Allowing the client to be built without knowing the deployment destination's domain name

---

Here's a patch in which I have made it so the client derives the ORCID Auth URI dynamically at run time:

```diff
$ git --no-pager diff webapp/client/src/common/UM/Common/ORCIDLogin.js
diff --git a/webapp/client/src/common/UM/Common/ORCIDLogin.js b/webapp/client/src/common/UM/Common/ORCIDLogin.js
index 74e68e9..007e3c8 100644
--- a/webapp/client/src/common/UM/Common/ORCIDLogin.js
+++ b/webapp/client/src/common/UM/Common/ORCIDLogin.js
@@ -8,6 +8,61 @@ import { popupWindow } from '../../util';
 import { socialLogin, cleanupMessages } from "../../../redux/actions/userActions";
 import { LoaderDialog, ConfirmDialogNoHeader, MessageDialog } from '../../Dialogs';
 
+/**
+ * Returns a local URI based upon the specified parameters; and the current protocol and domain.
+ *
+ * References:
+ * - https://developer.mozilla.org/en-US/docs/Web/API/Location/protocol
+ * - https://developer.mozilla.org/en-US/docs/Web/API/Location/hostname
+ * - https://developer.mozilla.org/en-US/docs/Web/API/Location/port
+ *
+ * @param path {string} The path portion of the URI, including the leading forward slash
+ * @return {string} The resulting local URI
+ */
+const makeLocalUri = (path = "/oauth") => {
+    const protocol = window.location.protocol; // e.g. "http:" or "https:"
+    const domain = window.location.hostname; // e.g. "www.example.com"
+    const port = window.location.port; // e.g. "8000" or "80" or ""
+    const portIfAny = port === "" ? "" : `:${port}`;
+    return `${protocol}//${domain}${portIfAny}${path}`;
+};
+
+/**
+ * Returns an ORCiD Auth URI based upon the specified parameters.
+ *
+ * References:
+ * - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent
+ *
+ * @param redirectUri {string} URI to which you want ORCiD to redirect the client after authenticating
+ * @param orcidClientId {string} ORCiD `client_id` obtained from the ORCiD Developer Tools web page
+ * @param nonceVal {string} Nonce value we can use to verify that the redirected client originated here
+ * @param orcidAuthBaseUri {string} ORCiD oAuth base URI
+ * @return {string} The resulting ORCiD Auth URI
+ */
+const makeOrcidAuthUri = (redirectUri, orcidClientId, nonceVal, orcidAuthBaseUri="https://orcid.org/oauth/authorize") => {
+    const sanitizedRedirectUri = encodeURIComponent(redirectUri);
+    const sanitizedOrcidClientId = encodeURIComponent(orcidClientId);
+    const sanitizedNonceVal = encodeURIComponent(nonceVal);
+    return `${orcidAuthBaseUri}?response_type=token&redirect_uri=${sanitizedRedirectUri}&client_id=${sanitizedOrcidClientId}&scope=openid&nonce=${sanitizedNonceVal}`;
+}
+
+/**
+ * Extracts and returns the specified query parameter value from the specified URI.
+ *
+ * References:
+ * - https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams
+ * - https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/get
+ *
+ * @param uri {string} The URI containing the query string containing the specified parameter
+ * @param paramName {string} Name of the parameter whose value you want to extract
+ * @return {string} The extracted parameter value
+ */
+const extractQueryParamVal = (uri, paramName = "client_id") => {
+    const url = new URL(uri);
+    const params = new URLSearchParams(url.search);
+    return params.get(paramName);
+};
+
 export function ORCIDLogin(props) {
     const dispatch = useDispatch();
     const errors = useSelector(state => state.errors);
@@ -43,8 +98,11 @@ export function ORCIDLogin(props) {
     }
 
     const getORCID = () => {
-        //ORICD config
-        const url = process.env.REACT_APP_ORCID_AUTH_URL;
+        // TODO: Store ORCiD client ID in its own environment variable (e.g. `process.env.REACT_APP_ORCID_CLIENT_ID`)
+        //       so the `extractQueryParamVal` function is not necessary.
+        const orcidClientId = extractQueryParamVal(process.env.REACT_APP_ORCID_AUTH_URL, "client_id");
+        const url = makeOrcidAuthUri(makeLocalUri("/oauth"), orcidClientId, "whatever");
+
         //console.log(url)
         //open ORCiD login page
         popupWindow(url, 'ORCiD', window, 600, 700);

```

Note: I wrote that patch to work with the client environment variable file as it exists today—so it doesn't really get rid of the problem. However, if that environment variable file were changed so that the currently-internal `ORCID_CLIENT_ID` environment variable were exposed to the build process as `REACT_APP_ORCID_CLIENT_ID` (which I do recommend), then the patch could undergo the following additional changes:

```diff
-
-/**
- * Extracts and returns the specified query parameter value from the specified URI.
- *
- * References:
- * - https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams
- * - https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/get
- *
- * @param uri {string} The URI containing the query string containing the specified parameter
- * @param paramName {string} Name of the parameter whose value you want to extract
- * @return {string} The extracted parameter value
- */
-const extractQueryParamVal = (uri, paramName = "client_id") => {
-    const url = new URL(uri);
-    const params = new URLSearchParams(url.search);
-    return params.get(paramName);
-};
-
```

```diff
-        // TODO: Store ORCiD client ID in its own environment variable (e.g. `process.env.REACT_APP_ORCID_CLIENT_ID`)
-        //       so the `extractQueryParamVal` function is not necessary.
-        const orcidClientId = extractQueryParamVal(process.env.REACT_APP_ORCID_AUTH_URL, "client_id");
+        const orcidClientId = process.env.REACT_APP_ORCID_CLIENT_ID;
         const url = makeOrcidAuthUri(makeLocalUri("/oauth"), orcidClientId, "whatever");
```

**Comment by eecavanna** [comment]
This is fixed in the following large commit, which is part of the project of containerizing the project.

e41b8d7c6e349efe53e806ea8823c5f10671d0c9

**Comment by eecavanna** [comment]
The fix for this was merged into `main` via PR #17
